### PR TITLE
Suppress the Jetty server version header

### PR DESCRIPTION
**Pull request by milosimpson**

### CODE REVIEW

**Comment by milosimpson** [comment]
For safety, the version information should be hidden / suppressed.

File: core/src/test/java/io/confluent/rest/SslTest.java [hunk 1]
```diff
@@ -293,6 +295,11 @@ private int makeGetRequest(String url, String clientKeystoreLocation, String cli
     CloseableHttpResponse response = null;
     try {
       response = httpclient.execute(httpget);
+
+      // Verify that the Jetty Server version header is not returned
```
Comment (big-andy-coates) [inline core/src/test/java/io/confluent/rest/SslTest.java]: This is not an ideal place to be adding such test code. This is executed for each test in `SslTest` and this check has nothing to do with SSL.

Can I suggest you add a test to `RestApiTest.java` that explicitly tests just this one thing? e.g.

```java
  @Test
  public void shouldNotExposeServerVersionInfo() {
    // Given:
    final Client client = ClientBuilder.newBuilder().build();
    try {

      // When:
      final Response res = client
          .target(REST_APP.getHttpListener())
          .path("/info")
          .request(MediaType.APPLICATION_JSON_TYPE)
          .get();

      // Then:
      assertThat(res.getStatusInfo().getStatusCode(), is(HttpStatus.SC_OK));
      final String version = res.getHeaderString("Server");
      assertThat(version, is(nullValue()));

    } finally {
      client.close();
    }
  }
```

File: core/src/main/java/io/confluent/rest/Application.java [hunk 2]
```diff
@@ -228,13 +230,21 @@ protected void doStop() throws Exception {
 
     List<URI> listeners = parseListeners(config.getList(RestConfig.LISTENERS_CONFIG),
             config.getInt(RestConfig.PORT_CONFIG), Arrays.asList("http", "https"), "http");
+
+    // Make an HttpConfiguration that turns off the server version header
+    HttpConfiguration httpConfiguration = new HttpConfiguration();
```
Comment (big-andy-coates) [inline core/src/main/java/io/confluent/rest/Application.java]: nit: final

File: core/src/main/java/io/confluent/rest/Application.java [hunk 3]
```diff
@@ -228,13 +230,21 @@ protected void doStop() throws Exception {
 
     List<URI> listeners = parseListeners(config.getList(RestConfig.LISTENERS_CONFIG),
             config.getInt(RestConfig.PORT_CONFIG), Arrays.asList("http", "https"), "http");
+
+    // Make an HttpConfiguration that turns off the server version header
```
Comment (big-andy-coates) [inline core/src/main/java/io/confluent/rest/Application.java]: nit: remove comment - it's unnecessary (and these things tend to get out of date).

File: core/src/test/java/io/confluent/rest/SslTest.java (hunk 1 above)
Comment (milosimpson) [reply core/src/test/java/io/confluent/rest/SslTest.java]: Want to make sure the header is gone for both http and https, and SslTest is the only one that stands up https.   To put this in it's own test class (there is not pre-existing RestApiTest.java) the ssl setup would need to be duped / refactored to be shared with another test class.

I'll add a dedicated test method for this to SslTest.

File: core/src/main/java/io/confluent/rest/Application.java (hunk 2 above)
Comment (milosimpson) [reply core/src/main/java/io/confluent/rest/Application.java]: done

File: core/src/test/java/io/confluent/rest/SslTest.java (hunk 1 above)
Comment (milosimpson) [reply core/src/test/java/io/confluent/rest/SslTest.java]: Duped and simplified the SslTest class, so there is now an ApiHeadersTest that tests both http and https.

File: core/src/main/java/io/confluent/rest/Application.java (hunk 3 above)
Comment (milosimpson) [reply core/src/main/java/io/confluent/rest/Application.java]: Done

File: core/src/test/java/io/confluent/rest/ApiHeadersTest.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+import org.apache.kafka.test.TestSslUtils.CertificateBuilder;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class ApiHeadersTest {
+
+
+  private static String httpUri = "http://localhost:8080";
```
Comment (big-andy-coates) [inline core/src/test/java/io/confluent/rest/ApiHeadersTest.java]: nit: `final` on these two

File: core/src/test/java/io/confluent/rest/ApiHeadersTest.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+  }
+
+  @AfterClass
+  public static void teardown() throws Exception {
+    if (app != null) {
+      app.stop();
+    }
+  }
+
+  private static void createKeystoreWithCert(File file, String alias, Map<String, X509Certificate> certs) throws Exception {
```
Comment (big-andy-coates) [inline core/src/test/java/io/confluent/rest/ApiHeadersTest.java]: nit: private statics to the bottom of the file.

File: core/src/test/java/io/confluent/rest/ApiHeadersTest.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+  private static String httpsUri = "https://localhost:8081";
+
+  private static final String SSL_PASSWORD = "test1234";
+
+  private static String clientKeystoreLocation;
+  private static TestApplication app;
+
+  @BeforeClass
+  public static void setUp() throws Exception {
+    File trustStore = File.createTempFile("ApiHeadersTest-truststore", ".jks");
```
Comment (big-andy-coates) [inline core/src/test/java/io/confluent/rest/ApiHeadersTest.java]: nit: `final` on all vars through out this file.

File: core/src/test/java/io/confluent/rest/ApiHeadersTest.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    HttpGet httpget = new HttpGet(httpUri +  "/test/endpoint");
+
+    try ( CloseableHttpClient httpclient = HttpClients.createDefault();
+        CloseableHttpResponse response = httpclient.execute(httpget) ) {
+
+      int statusCode = response.getStatusLine().getStatusCode();
+      Assert.assertEquals(200, statusCode);
+
+      Header jettyServerVersionHeader = response.getFirstHeader( "Server" );
+      assertNull(jettyServerVersionHeader);
```
Comment (big-andy-coates) [inline core/src/test/java/io/confluent/rest/ApiHeadersTest.java]: nit: use Hamcrest's `assertThat` for all new code.

```suggestion
      assertThat(jettyServerVersionHeader, is(nullValue()));
```

File: core/src/test/java/io/confluent/rest/ApiHeadersTest.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+        null, SSLConnectionSocketFactory.getDefaultHostnameVerifier());
+
+    try ( CloseableHttpClient httpclient = HttpClients.custom().setSSLSocketFactory(sslSf).build();
+        CloseableHttpResponse response = httpclient.execute(httpget) ) {
+
+      int statusCode = response.getStatusLine().getStatusCode();
+      Assert.assertEquals(200, statusCode);
+
+      Header jettyServerVersionHeader = response.getFirstHeader( "Server" );
+      assertNull(jettyServerVersionHeader);
```
Comment (big-andy-coates) [inline core/src/test/java/io/confluent/rest/ApiHeadersTest.java]: nit: use Hamcrest's `assertThat` for all new code.

```suggestion
      assertThat(jettyServerVersionHeader, is(nullValue()));
```

File: core/src/test/java/io/confluent/rest/ApiHeadersTest.java [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    }
+  }
+
+
+  private static class TestApplication extends Application<TestRestConfig> {
```
Comment (big-andy-coates) [inline core/src/test/java/io/confluent/rest/ApiHeadersTest.java]: Is there a reason you've not just used `TestKsqlRestApp`, with `ServerKeyStore`, and `ClientTrustStore` and then tried to access `\info`??

File: core/src/test/java/io/confluent/rest/ApiHeadersTest.java (hunk 4 above)
Comment (milosimpson) [reply core/src/test/java/io/confluent/rest/ApiHeadersTest.java]: done

File: core/src/test/java/io/confluent/rest/ApiHeadersTest.java (hunk 5 above)
Comment (milosimpson) [reply core/src/test/java/io/confluent/rest/ApiHeadersTest.java]: done

File: core/src/test/java/io/confluent/rest/ApiHeadersTest.java (hunk 6 above)
Comment (milosimpson) [reply core/src/test/java/io/confluent/rest/ApiHeadersTest.java]: done

File: core/src/test/java/io/confluent/rest/ApiHeadersTest.java (hunk 7 above)
Comment (milosimpson) [reply core/src/test/java/io/confluent/rest/ApiHeadersTest.java]: done

File: core/src/test/java/io/confluent/rest/ApiHeadersTest.java (hunk 8 above)
Comment (milosimpson) [reply core/src/test/java/io/confluent/rest/ApiHeadersTest.java]: Done

File: core/src/test/java/io/confluent/rest/ApiHeadersTest.java (hunk 9 above)
Comment (milosimpson) [reply core/src/test/java/io/confluent/rest/ApiHeadersTest.java]: So, there is no `TestKsqlRestApp` in the rest-utils project.   

Also, I looked at all things that `extend Application`, and none of them look like good candidates to reuse.

**Comment by big-andy-coates** [comment]
Thanks @milosimpson  LGTM

File: core/src/test/java/io/confluent/rest/ApiHeadersTest.java (hunk 9 above)
Comment (big-andy-coates) [reply core/src/test/java/io/confluent/rest/ApiHeadersTest.java]: Sorry, getting my repo's mixed up!

**Pull request merged**